File: tools/ao-export/export_grh.py
"""
Exporta uno o más Grh de graficos.ini a PNG recortados.
Uso:
  python export_grh.py --grh 63 --recursos RUTA/Recursos --out salida.png
  python export_grh.py --grh 100 --recursos RUTA/Recursos --out-dir carpeta/
"""

from __future__ import annotations

import argparse
import re
from dataclasses import dataclass
from pathlib import Path

from PIL import Image

from export_bmp_to_png import TRANSPARENT_RGB, TOLERANCE

GRH_LINE = re.compile(r"^Grh(\d+)=(.+)$", re.IGNORECASE)


@dataclass
class GrhFrame:
    grh_id: int
    file_num: int
    x: int
    y: int
    width: int
    height: int


def load_graficos_ini(path: Path) -> dict[int, str]:
    grhs: dict[int, str] = {}
    with path.open(encoding="latin-1", errors="replace") as f:
        for line in f:
            line = line.strip()
            match = GRH_LINE.match(line)
            if match:
                grhs[int(match.group(1))] = match.group(2)
    return grhs


def parse_static_grh(grh_id: int, value: str) -> list[GrhFrame]:
    """Grh estático: 1-archivo-x-y-ancho-alto[-descripcion]"""
    parts = value.split("-")
    if len(parts) < 6:
        raise ValueError(f"Grh{grh_id} formato no reconocido: {value}")

    frame_count = int(parts[0])
    if frame_count != 1:
        raise ValueError(
            f"Grh{grh_id} es animación ({frame_count} frames). "
            f"Usá --grh con el id de animación o exportá cada frame por separado."
        )

    return [
        GrhFrame(
            grh_id=grh_id,
            file_num=int(parts[1]),
            x=int(parts[2]),
            y=int(parts[3]),
            width=int(parts[4]),
            height=int(parts[5]),
        )
    ]


def parse_animation_grh(grh_id: int, value: str, all_grhs: dict[int, str]) -> list[GrhFrame]:
    """Animación: N-grh1-grh2-...-velocidad"""
    parts = value.split("-")
    frame_count = int(parts[0])
    frame_ids = [int(p) for p in parts[1 : 1 + frame_count]]
    frames: list[GrhFrame] = []

    for fid in frame_ids:
        if fid not in all_grhs:
            raise KeyError(f"Frame Grh{fid} no encontrado (referenciado por Grh{grh_id})")
        frames.extend(parse_static_grh(fid, all_grhs[fid]))

    return frames


def resolve_frames(grh_id: int, all_grhs: dict[int, str]) -> list[GrhFrame]:
    if grh_id not in all_grhs:
        raise KeyError(f"Grh{grh_id} no existe en graficos.ini")

    value = all_grhs[grh_id]
    parts = value.split("-")
    frame_count = int(parts[0])

    if frame_count == 1:
        return parse_static_grh(grh_id, value)
    return parse_animation_grh(grh_id, value, all_grhs)


def find_bmp(bmp_dir: Path, file_num: int) -> Path:
    for ext in (".bmp", ".BMP"):
        candidate = bmp_dir / f"{file_num}{ext}"
        if candidate.exists():
            return candidate
    raise FileNotFoundError(f"No se encontró {file_num}.bmp en {bmp_dir}")


def crop_frame(bmp_path: Path, frame: GrhFrame) -> Image.Image:
    img = Image.open(bmp_path).convert("RGBA")
    cropped = img.crop(
        (frame.x, frame.y, frame.x + frame.width, frame.y + frame.height)
    )
    pixels = cropped.load()
    w, h = cropped.size

    for y in range(h):
        for x in range(w):
            r, g, b, a = pixels[x, y]
            if (
                abs(r - TRANSPARENT_RGB[0]) <= TOLERANCE
                and abs(g - TRANSPARENT_RGB[1]) <= TOLERANCE
                and abs(b - TRANSPARENT_RGB[2]) <= TOLERANCE
            ):
                pixels[x, y] = (0, 0, 0, 0)

    return cropped


def main() -> None:
    parser = argparse.ArgumentParser(description="Exportar Grh de AO a PNG")
    parser.add_argument("--grh", type=int, required=True, help="Número de Grh a exportar")
    parser.add_argument(
        "--recursos",
        required=True,
        help="Carpeta Recursos del clone (contiene init/graficos.ini y Graficos/bmp)",
    )
    parser.add_argument("--out", help="Archivo PNG de salida (un solo frame)")
    parser.add_argument("--out-dir", help="Carpeta de salida (varios frames)")
    args = parser.parse_args()

    recursos = Path(args.recursos)
    ini_path = recursos / "init" / "graficos.ini"
    bmp_dir = recursos / "Graficos" / "bmp"

    if not ini_path.exists():
        print(f"ERROR: No existe {ini_path}")
        return

    if not bmp_dir.is_dir():
        print(f"ERROR: No existe {bmp_dir}")
        print("Copiá los BMP desde la instalación de AO — ver README.md")
        return

    all_grhs = load_graficos_ini(ini_path)
    frames = resolve_frames(args.grh, all_grhs)

    if args.out_dir:
        out_dir = Path(args.out_dir)
        out_dir.mkdir(parents=True, exist_ok=True)
        for i, frame in enumerate(frames):
            bmp_path = find_bmp(bmp_dir, frame.file_num)
            img = crop_frame(bmp_path, frame)
            out_path = out_dir / f"grh{frame.grh_id}_{i}.png"
            img.save(out_path, "PNG")
            print(f"Guardado: {out_path}")
        return

    if len(frames) > 1:
        print("Este Grh tiene varios frames. Usá --out-dir en lugar de --out.")
        return

    out = Path(args.out or f"grh{args.grh}.png")
    frame = frames[0]
    bmp_path = find_bmp(bmp_dir, frame.file_num)
    img = crop_frame(bmp_path, frame)
    out.parent.mkdir(parents=True, exist_ok=True)
    img.save(out, "PNG")
    print(f"Guardado: {out} ({frame.width}x{frame.height} desde {bmp_path.name})")


if __name__ == "__main__":
    main()

File: tools/ao-export/export-grh.mjs
/**
 * Exporta un Grh de graficos.ini a PNG
 *
 * npm run grh -- --grh 63 --recursos "C:\...\RecursosAO\Recursos" --out "salida.png"
 */

import fs from "fs";
import path from "path";
import bmp from "bmp-js";
import { PNG } from "pngjs";

const TOLERANCE = 8;

function parseArgs() {
  const args = process.argv.slice(2);
  const opts = {};
  const positional = [];

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--grh") opts.grh = Number(args[++i]);
    else if (args[i] === "--recursos") opts.recursos = args[++i];
    else if (args[i] === "--out") opts.out = args[++i];
    else if (args[i] === "--out-dir") opts.outDir = args[++i];
    else if (!args[i].startsWith("--")) positional.push(args[i]);
  }

  if (!opts.recursos && positional[0]) opts.recursos = positional[0];
  if (!opts.out && positional[1]) opts.out = positional[1];

  return opts;
}

function isTransparent(r, g, b) {
  return (
    Math.abs(r) <= TOLERANCE &&
    Math.abs(g) <= TOLERANCE &&
    Math.abs(b) <= TOLERANCE
  );
}

function loadGrhMap(iniPath) {
  const text = fs.readFileSync(iniPath, "latin1");
  const map = new Map();
  for (const line of text.split(/\r?\n/)) {
    const m = line.trim().match(/^Grh(\d+)=(.+)$/i);
    if (m) map.set(Number(m[1]), m[2]);
  }
  return map;
}

function resolveStatic(grhId, value, grhs) {
  const parts = value.split("-");
  const frameCount = Number(parts[0]);
  if (frameCount !== 1) {
    const frameIds = parts.slice(1, 1 + frameCount).map(Number);
    return frameIds.flatMap((id) => resolveStatic(id, grhs.get(id), grhs));
  }
  return [
    {
      grhId,
      fileNum: Number(parts[1]),
      x: Number(parts[2]),
      y: Number(parts[3]),
      w: Number(parts[4]),
      h: Number(parts[5]),
    },
  ];
}

function findBmp(bmpDir, fileNum) {
  for (const ext of [".bmp", ".BMP"]) {
    const p = path.join(bmpDir, `${fileNum}${ext}`);
    if (fs.existsSync(p)) return p;
  }
  throw new Error(`No se encontró ${fileNum}.bmp en ${bmpDir}`);
}

function cropFrame(bmpPath, frame) {
  const { data, width } = bmp.decode(fs.readFileSync(bmpPath));
  const png = new PNG({ width: frame.w, height: frame.h });

  for (let y = 0; y < frame.h; y++) {
    for (let x = 0; x < frame.w; x++) {
      const si = (width * (frame.y + y) + (frame.x + x)) * 4;
      const di = (frame.w * y + x) << 2;
      const r = data[si];
      const g = data[si + 1];
      const b = data[si + 2];
      png.data[di] = r;
      png.data[di + 1] = g;
      png.data[di + 2] = b;
      png.data[di + 3] = isTransparent(r, g, b) ? 0 : 255;
    }
  }
  return png;
}

const { grh, recursos, out, outDir } = parseArgs();

if (!grh || !recursos) {
  console.log("Uso:");
  console.log(
    '  npm run grh -- --grh 63 --recursos "RUTA\\RecursosAO\\Recursos" --out "grh63.png"'
  );
  process.exit(1);
}

const iniPath = path.join(recursos, "init", "graficos.ini");
const bmpDir = path.join(recursos, "Graficos", "bmp");

if (!fs.existsSync(iniPath)) {
  console.error(`ERROR: No existe ${iniPath}`);
  process.exit(1);
}
if (!fs.existsSync(bmpDir)) {
  console.error(`ERROR: No existe ${bmpDir}`);
  console.error("Copiá los BMP desde AO (Steam) — ver README.md");
  process.exit(1);
}

const grhs = loadGrhMap(iniPath);
if (!grhs.has(grh)) {
  console.error(`ERROR: Grh${grh} no existe en graficos.ini`);
  process.exit(1);
}

const frames = resolveStatic(grh, grhs.get(grh), grhs);

if (outDir) {
  fs.mkdirSync(outDir, { recursive: true });
  for (let i = 0; i < frames.length; i++) {
    const frame = frames[i];
    const bmpPath = findBmp(bmpDir, frame.fileNum);
    const png = cropFrame(bmpPath, frame);
    const dst = path.join(outDir, `grh${frame.grhId}_${i}.png`);
    fs.writeFileSync(dst, PNG.sync.write(png));
    console.log(`Guardado: ${dst}`);
  }
} else if (frames.length > 1) {
  console.error("Este Grh tiene varios frames. Usá --out-dir.");
  process.exit(1);
} else {
  const frame = frames[0];
  const bmpPath = findBmp(bmpDir, frame.fileNum);
  const png = cropFrame(bmpPath, frame);
  const dst = out || `grh${grh}.png`;
  fs.mkdirSync(path.dirname(path.resolve(dst)), { recursive: true });
  fs.writeFileSync(dst, PNG.sync.write(png));
  console.log(`Guardado: ${dst} (${frame.w}x${frame.h} desde ${path.basename(bmpPath)})`);
}
